### ISSUE TITLE
train failed on multiGPUs 

### ISSUE THREAD

**Issue by JoeyHeisenberg** [Tomiinek/Multilingual_Text_to_Speech#6]
It work fine on single GPU, but failed on 3 GPUs

only change "max_gpus" to 3, is there anything I should change?

CUDA_VISIBLE_DEVICES=0,1,2 PYTHONIOENCODING=utf-8 nohup python -u train.py --hyper_parameters generated_training > log.file 2>&1 &

![image](https://user-images.githubusercontent.com/33407667/84028985-44bad300-a9c4-11ea-8f00-50d1b392d183.png)

**Comment by Tomiinek** [comment]
Hi, could you please provide the `.json` file with parameters you are using?

There is a mismatch in shapes. You are using the generated encoder, so the `batch_size` must be divisible by the number of lanugages you are training on times the number of GPUs. So let's say you are trianing on CSS10 with 10 languages and you want to use 3 GPUs, than set the `batch_size` parameter to 30, 60, 90, ... 

Hope it helps :grin:

**Comment by JoeyHeisenberg** [comment]
The above results are actually with the batch_size setting to 60

I tried to set to 180, but it still failed

`Original Traceback (most recent call last):
  File "/data/glusterfs_speech_tts/public_data/11104653/tools/miniconda3/envs/envMTS/lib/python3.6/site-packages/torch/nn/parallel/parallel_apply.py", line 60, in _worker
    output = module(*input, **kwargs)
  File "/data/glusterfs_speech_tts/public_data/11104653/tools/miniconda3/envs/envMTS/lib/python3.6/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "/data/glusterfs_speech_tts/public_data/11104653/multiLingual_voice_cloning/Multilingual_Text_to_Speech/modules/tacotron2.py", line 364, in forward
    encoded = self._encoder(embedded, text_length, languages)
  File "/data/glusterfs_speech_tts/public_data/11104653/tools/miniconda3/envs/envMTS/lib/python3.6/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "/data/glusterfs_speech_tts/public_data/11104653/multiLingual_voice_cloning/Multilingual_Text_to_Speech/modules/encoder.py", line 208, in forward
    x = x.reshape(bs // self._groups, self._groups * self._input_dim, -1)
RuntimeError: shape '[3, 5120, -1]' is invalid for input of size 2994176`

HERE IS THE .json
`{
        "balanced_sampling": true,
        "batch_size": 180,
        "case_sensitive": false,
        "characters": " abcdefghijklmnopqrstuvwxyzçèéßäöōǎǐíǒàáǔüèéìūòóùúāēěīâêôûñőűабвгдежзийклмнопрстуфхцчшщъыьэюяёάέήίαβγδεζηθικλμνξοπρςíστυφχψωόύώ",
        "checkpoint_each_epochs": 5,
        "dataset": "css10",
        "encoder_dimension": 256,
        "encoder_type": "generated",
        "epochs": 300,
        "generator_bottleneck_dim": 8,
        "generator_dim": 20,
        "languages": ["german", "french", "hungarian", "chinese", "spanish", "dutch", "finnish", "russian", "japanese", "greek"],
        "language_embedding_dimension": 32,
        "learning_rate": 0.001,
        "learning_rate_decay_each": 10000,
        "learning_rate_decay_start": 10000,
        "multi_language": true,
        "perfect_sampling": true,
        "predict_linear": false,
        "version": "GENERATED-TRAINING"
}`

**Comment by Tomiinek** [comment]
Oh, ok. Still can't figure out where is the problem :cry:  

Could you please place `print(x.shape)` to the line 197 of `encoder.py` and send me the output? I cannot test it myself now.

If you want to make it working immediately, change line 229 in `train.py` from:
```
eval_sampler = PerfectBatchSampler(dataset.dev, hp.languages, hp.batch_size, data_parallel_devices=dp_devices, shuffle=False) 
```
to
```
eval_sampler = PerfectBatchSampler(dataset.dev, hp.languages, hp.batch_size, data_parallel_devices=dp_devices, shuffle=False, drop_last=True)
``` 
and it should work.

**Comment by JoeyHeisenberg** [comment]
Here is the result

`eval
x.shape:  torch.Size([60, 173, 512])
x.shape:  torch.Size([60, 173, 512])
x.shape:  torch.Size([60, 173, 512])
x.shape:  torch.Size([60, 173, 512])
x.shape:  torch.Size([60, 173, 512])
x.shape:  torch.Size([60, 173, 512])
x.shape:  torch.Size([60, 165, 512])
x.shape:  torch.Size([60, 165, 512])
x.shape:  torch.Size([60, 165, 512])
x.shape:  torch.Size([60, 165, 512])
x.shape:  torch.Size([60, 165, 512])
x.shape:  torch.Size([60, 165, 512])
x.shape:  torch.Size([60, 168, 512])
x.shape:  torch.Size([60, 168, 512])
x.shape:  torch.Size([60, 168, 512])
x.shape:  torch.Size([60, 168, 512])
x.shape:  torch.Size([60, 168, 512])
x.shape:  torch.Size([60, 168, 512])
x.shape:  torch.Size([34, 172, 512])
x.shape:  torch.Size([34, 172, 512])
x.shape:  torch.Size([32, 172, 512])
Traceback (most recent call last):
  File "train.py", line 302, in <module>
    eval_loss = evaluate(epoch, eval_data, model, criterion)   
  File "train.py", line 124, in evaluate
    post_pred, pre_pred, stop_pred, alignment, spkrs_pred, enc_output = model(src, src_len, trg_mel, trg_len, spkrs, langs, 1.0)
  File "/data/glusterfs_speech_tts/public_data/11104653/tools/miniconda3/envs/envMTS/lib/python3.6/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "/data/glusterfs_speech_tts/public_data/11104653/tools/miniconda3/envs/envMTS/lib/python3.6/site-packages/torch/nn/parallel/data_parallel.py", line 152, in forward
    outputs = self.parallel_apply(replicas, inputs, kwargs)
  File "/data/glusterfs_speech_tts/public_data/11104653/tools/miniconda3/envs/envMTS/lib/python3.6/site-packages/torch/nn/parallel/data_parallel.py", line 162, in parallel_apply
    return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
  File "/data/glusterfs_speech_tts/public_data/11104653/tools/miniconda3/envs/envMTS/lib/python3.6/site-packages/torch/nn/parallel/parallel_apply.py", line 85, in parallel_apply
    output.reraise()
  File "/data/glusterfs_speech_tts/public_data/11104653/tools/miniconda3/envs/envMTS/lib/python3.6/site-packages/torch/_utils.py", line 385, in reraise
    raise self.exc_type(msg)
RuntimeError: Caught RuntimeError in replica 0 on device 0.
Original Traceback (most recent call last):
  File "/data/glusterfs_speech_tts/public_data/11104653/tools/miniconda3/envs/envMTS/lib/python3.6/site-packages/torch/nn/parallel/parallel_apply.py", line 60, in _worker
    output = module(*input, **kwargs)
  File "/data/glusterfs_speech_tts/public_data/11104653/tools/miniconda3/envs/envMTS/lib/python3.6/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "/data/glusterfs_speech_tts/public_data/11104653/multiLingual_voice_cloning/Multilingual_Text_to_Speech/modules/tacotron2.py", line 364, in forward
    encoded = self._encoder(embedded, text_length, languages)
  File "/data/glusterfs_speech_tts/public_data/11104653/tools/miniconda3/envs/envMTS/lib/python3.6/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "/data/glusterfs_speech_tts/public_data/11104653/multiLingual_voice_cloning/Multilingual_Text_to_Speech/modules/encoder.py", line 208, in forward
    x = x.reshape(bs // self._groups, self._groups * self._input_dim, -1)
RuntimeError: shape '[3, 5120, -1]' is invalid for input of size 2994176
`

**Comment by Tomiinek** [comment]
Should be solved by 0401a6d. Thank you!
